### ISSUE TITLE
Implement __eq__

### ISSUE THREAD

**Issue by antonio-one** [antonio-one/table_loader#13]
3. Implement __eq__ in Table to have the different posibilities considered:
  - Test case Comparison between the same table in local and gcs storage to yield equality true
  - Test case Comparison between the same table in local and bigquery to yield equality true
  - Test case Comparison between the same table in gcs and bigquery to yield equiality true
  - Opposite test cases (one attribute yields difference)

**Comment by antonio-one** [comment]
### Notes
In order to assess equality the following must be true:
    - both pbjects must be Table(s) and their grid_id(s) must be equal
    - at least the same one attribute other than the grid_id must exist in both objects and be of equal value
    - if more than one same attribute exist in both objects, they must all be of equal value
This equation is not aware of the three Table states.

### Initial Test Results
```
table_loader/tests/unit/test_model.py::test_schema_is_set PASSED
table_loader/tests/unit/test_model.py::test_local_and_gcs_tables_are_equal PASSED
table_loader/tests/unit/test_model.py::test_local_and_gcs_tables_are_not_equal PASSED
table_loader/tests/unit/test_model.py::test_local_and_bq_tables_are_equal PASSED
table_loader/tests/unit/test_model.py::test_local_and_bq_tables_are_not_equal PASSED
table_loader/tests/unit/test_model.py::test_gcs_and_bq_tables_are_equal FAILED
table_loader/tests/unit/test_model.py::test_gcs_and_bq_tables_are_not_equal FAILED
```
Assuming we don't want to download anything, gcs and bq objects have nothing in common.

### GCS/BQ Comparison Options
In the code excerpt below the three options we could combine are:

1. payload_hash
2. schema
3. rows

```
    gcs_table = model.Table()
    gcs_table.grid_id = grid_id
    gcs_table.payload_hash = tuv.PAYLOAD_2_HASH

    bq_table = model.Table()
    bq_table.grid_id = grid_id
    bq_table.schema = tuv.JSON_SCHEMA_1
    bq_table.rows = tuv.ROWS_2
```

In order to assert equality it would be recommended to use at least 2 if not all 3 attributes:

1. Download schema from GCA bucket <=> Get schema via the BQ api
2. Get payload_hash via the GCS api <=> Calculate a payload_hash by querying the BQ table (one could save a "stale" hash from the upload as a label but that would remain unchanged in the event of a manual BQ change)
3. Save the number of rows in a meta file <=> Get number of rows via the BQ api (from the table meta)

**Comment by txomon** [comment]
Excellent analysis, thanks for your in detail explanation.

> We don't compare GCS to BQ at all and merely rely on the comparisons of local <=> GCS and local <=> BQ. If either of them are out of sync we assume that a sync needs to be executed.

I believe this case wouldn't allow us to know whether the GBQ tables have been modified or not.

The way I see it is:

 * The only source of truth we have from GBQ is modification time, and this one can only be set in GCS (https://cloud.google.com/storage/docs/metadata#custom-time).

 * Between local and GCS, we can only rely on the fact that the same file names are in both places and their hashes, as we don't want to transfer data to assess diferences.

Please correct if these two assumptions are not correct, as the following builds on top of it.

I believe the sync should happen between GCS and GBQ, using modification time between the two, and between GCS and Local storage using file hashing.

Hence, we could check that the objects to be compared have at least either modification times or hashings, preferring hashing over modification times, so that would mean our Table object has a notion of what files define it.

How does it sound to you?

**Comment by txomon** [comment]
Let's move the conversation from the implementantion PR to the issue that documents backend attributes https://github.com/antonio-one/table_loader/issues/12

**Comment by antonio-one** [comment]
There are [multiple truths in BQ](https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#resource-table), some of which could be relevant to this.

How do we know if BQ has he right data without cost originating from downloads?

We could maintain the below labels every time we update the table:
  "numRows": string,
  "creationTime": string,
  "lastModifiedTime": string
... and compare them to the ones maintained by BQ itself. If a change has occurred we sync.

So in summary, if all the below are true, we don't need to sync:
```
local.payload_hash == gcs.payload_hash

local.rows == bq.rows
local.schema == bq.schema

bq.label.numRows == bq.rows
bq.label.creationTime == bq.creationTime
bq.label.lastModifiedTime == bq.lastModifiedTime
```
